californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -321,8 +321,17 @@ public void onTimeout() {
 					}
 
 					@Override
-					public void onResponse(final Response resp) {
-						notificationListener.onNotification(request, resp);
+					public void onResponse(Response response) {
+						if (!response.getOptions().hasObserve()) {
```

Comment:
I had the comment in the line below. May be your right to add it before the if loop.